WooCommerce/Classes/ViewRelated/Feature Highlight/Tooltip.swift

```diff
@@ -0,0 +1,420 @@
+        arrowPath.addLine(to: CGPoint(x: arrowOriginX, y: arrowTipY))
+        // Add curve until `arrowWidth/2 + 1` (2 points of curve for a rounded arrow tip).
+        arrowPath.addQuadCurve(
+            to: CGPoint(x: arrowOriginX + 2, y: arrowTipY),
+            controlPoint: CGPoint(x: Self.arrowWidth/2, y: arrowTipYControl)
+        )
+        // Draw down to 20.
+        arrowPath.addLine(to: CGPoint(x: Self.arrowWidth, y: 0))
+        arrowPath.close()
+
+        arrowShapeLayer = CAShapeLayer()
+        guard let arrowShapeLayer = arrowShapeLayer else {
+            return
+        }
+
+        arrowShapeLayer.path = arrowPath.cgPath
+
+        arrowShapeLayer.strokeColor = Constants.invertedTooltipBackgroundColor.cgColor
+        arrowShapeLayer.fillColor = Constants.invertedTooltipBackgroundColor.cgColor
+        arrowShapeLayer.lineWidth = 4.0
+
+        arrowShapeLayer.position = CGPoint(x: offsetX - Self.arrowWidth/2, y: offsetY)
+
+        containerView.layer.addSublayer(arrowShapeLayer)
+    }
+
+    func size() -> CGSize {
+        CGSize(
+            width: Self.width(
+                title: titleLabel.text,
+                message: message,
+                primaryButtonTitle: primaryButton.titleLabel?.text,
+                secondaryButtonTitle: secondaryButton.titleLabel?.text
+            ),
+            height: Self.height(
+                withTitle: title,
+                message: message
+            )
+        )
+    }
+
+    func copy() -> Tooltip {
+        let copyTooltip = Tooltip()
+        copyTooltip.title = title
+        copyTooltip.message = message
+        copyTooltip.primaryButtonTitle = primaryButtonTitle
+        copyTooltip.secondaryButtonTitle = secondaryButtonTitle
+        copyTooltip.dismissalAction = dismissalAction
+        copyTooltip.secondaryButtonAction = secondaryButtonAction
+        copyTooltip.shouldPrefixLeadingIcon = shouldPrefixLeadingIcon
+        copyTooltip.buttonAlignment = buttonAlignment
+        return copyTooltip
+    }
+
+    private func commonInit() {
+        backgroundColor = .clear
+
+        setUpContainerView()
+        setUpConstraints()
+        addShadow()
+        isAccessibilityElement = true
+    }
+
+    private func setUpContainerView() {
+        containerView.backgroundColor = Constants.invertedTooltipBackgroundColor
+        containerView.layer.cornerRadius = Constants.cornerRadius
+        containerView.translatesAutoresizingMaskIntoConstraints = false
+        addSubview(containerView)
+
+        containerTopConstraint = containerView.topAnchor.constraint(equalTo: topAnchor)
+        containerBottomConstraint = bottomAnchor.constraint(equalTo: containerView.bottomAnchor)
+
+        NSLayoutConstraint.activate([
+            containerTopConstraint!,
+            containerView.leadingAnchor.constraint(equalTo: leadingAnchor),
+            trailingAnchor.constraint(equalTo: containerView.trailingAnchor),
+            containerBottomConstraint!
```

Comment:
we generally don't force unwrap in the code freeze, how about keeping the constraints as local let constants before assigning to optional properties? or try `.compactMap { $0 }` with the `[NSLayoutConstraint]`?

Reply:
Fixed by using `compactMap` in https://github.com/woocommerce/woocommerce-ios/pull/10005/commits/beac2227ee06013d17a0326f802ce53ac0eb87fe